### PR TITLE
New VDOM Rendering Engine

### LINKED ISSUE #54: New Dojo Vdom Rendering Engine
**Enhancement**

The existing `vdom` implementation is becoming difficult to maintain, enhance and performance tune due to starting with maquette's vdom and integrating dojo's widget concepts into the existing code base. To enable dojo to continue enhancing the core renderer's functionality and flexibility, a ground up re-design is required. The re-design will provide a backwards compatible solution that ensures both widgets and virtual dom nodes are equal citizens within the engine and enable enhancements such as scheduling to be addressed.

### Implementation Tracking 

 * [x] Basic widget and node diff'ing and rendering
 * [x] Basic event handlers
 * [x] Basic Render scheduling
 * [x] Support returning arrays
 * [x] Insert before calculation
 * [x] Sub tree rendering and depth ordering
 * [x] Registry support
 * [x] Support deferred properties
 * [x] Auto instance binding for widget and node functions
 * [x] SVG Support
 * [x] Support changing event handlers across renders
 * [x] DOM Merging Support
 * [x] Sync Mode
 * [x] Node operations (focus, blur, click, scrollIntoView)
 * [x] Support `dom` pragma
 * [x] Lifecycle hooks (onAttach, onDetach)
 * [ ] Advanced Render scheduling
 * [x] Exit, enter and ~update~ animation support
 * [x] No Performance Regressions versus existing vdom
 * [x] Passing Unit Tests

### PR DESCRIPTION
**Pull request by agubler**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code has been formatted with [`prettier`](https://prettier.io/) as per the [readme code style guidelines](./../#code-style)
* [x] Unit or Functional tests are included in the PR

**Description:**

A new implementation of the VDOM rendering engine. 

* Resolves performance regressions introduced by fixes for supporting inserting nodes in the correct position when a widget returns an array of `DNode`s.
* Unrolls the recursive loop used in the previous VDOM implementation to processing a queue of instructions for both the vdom side and the DOM commit.
* Provides a maintainable platform for future feature and performance enhancements
* Feature parity with the existing VDOM implementation
* Removes the requirement for the `ProjectorMixin` which reduces indirection and ultimately bundle size.
* Provides a shim `ProjectorMixin` to ensure backwards compatibility.
* Automatically detects whether a merge is required, without requiring a separate API.

#### Performance benchmarks: New vdom (left) vs the existing implementation (right)

interactive_results

#### Basic Usage

```ts
import { renderer } from '@dojo/framework/widget-core/vdom';
import { w } from '@dojo/framework/widget-core/d';
import App from './MyApp';

const r = render(() => w(MyApp, {});
r.mount();
```

#### Implementation Tracking 

 * [x] Basic widget and node diff'ing and rendering
 * [x] Basic event handlers
 * [x] Basic Render scheduling
 * [x] Support returning arrays
 * [x] Insert before calculation
 * [x] Sub tree rendering and depth ordering
 * [x] Registry support
 * [x] Support deferred properties
 * [x] Auto instance binding for widget and node functions
 * [x] SVG Support
 * [x] Support changing event handlers across renders
 * [x] DOM Merging Support
 * [x] Sync Mode
 * [x] Node operations (focus, blur, click, scrollIntoView)
 * [x] Support `dom` pragma
 * [x] Lifecycle hooks (onAttach, onDetach)
 * [ ] Advanced Render scheduling **Not being included as part of these changes**
 * [x] Exit, enter and ~update~ animation support **Update animation support has been removed**
 * [x] No Performance Regressions versus existing vdom (improvements)
 * [x] Custom element support
 * [x] Passing Unit Tests
 * [x] README updates
 * [x] Removed console logs for missing keys, issue raised to address general debugging before V4.

Resolves #54

### CODE REVIEW

File: src/widget-core/WidgetBase.ts [hunk 1]
```diff
@@ -261,19 +256,51 @@ export class WidgetBase<P = WidgetProperties, C extends DNode = DNode> implement
...
 	}
 
-	public __render__(): DNode | DNode[] {
-		const instanceData = widgetInstanceMap.get(this)!;
-		instanceData.dirty = false;
+	private _filterAndConvert(nodes: DNode[]): (WNode | VNode)[];
+	private _filterAndConvert(nodes: DNode): WNode | VNode;
+	private _filterAndConvert(nodes: DNode | DNode[]): (WNode | VNode) | (WNode | VNode)[];
+	private _filterAndConvert(nodes: DNode | DNode[]): (WNode | VNode) | (WNode | VNode)[] {
+		const isArray = Array.isArray(nodes);
```
Comment (JamesMilnerUK) [inline src/widget-core/WidgetBase.ts]: I would probably rename this nodesIsArray so this is clearer

File: src/widget-core/WidgetBase.ts [hunk 2]
```diff
@@ -261,19 +256,51 @@ export class WidgetBase<P = WidgetProperties, C extends DNode = DNode> implement
...
+		const filteredNodes = Array.isArray(nodes) ? nodes : [nodes];
```
Comment (JamesMilnerUK) [inline src/widget-core/WidgetBase.ts]: Tiny one; you could just do `isArray` rather `Array.isArray(nodes)` again

File: src/widget-core/WidgetBase.ts [hunk 3]
```diff
@@ -261,19 +256,51 @@ export class WidgetBase<P = WidgetProperties, C extends DNode = DNode> implement
...
+			}
+			if (isVNode(node) && node.deferredPropertiesCallback) {
+				const properties = node.deferredPropertiesCallback(false);
+				node.originalProperties = node.properties;
+				node.properties = { ...properties, ...node.properties };
+			}
+			node.bind = this;
+			convertedNodes.push(node);
+			if (node.children && node.children.length) {
+				node.children = this._filterAndConvert(node.children);
```
[the remark this answers is not in the document]
Comment (agubler) [reply src/widget-core/WidgetBase.ts]: I suspect that the difference is negligible, certainly not shown up as a bottleneck in any of the performance analysis I've done.

File: tests/widget-core/unit/d.ts [hunk 4]
```diff
@@ -152,51 +151,51 @@ registerSuite('d', {
...
-			assert.deepEqual(vnode.properties, {});
-			assert.deepEqual(vnode.attributes, {});
-			assert.deepEqual(vnode.children, undefined);
-			assert.strictEqual(vnode.diffType, 'none');
-		},
+	// dom: {
+	// 	'creates VNode with the DOM node attached, associated tagname and default diff type'() {
+	// 		const div = document.createElement('div');
+	// 		const vnode = dom(
+	// 			{
```
Comment (JamesMilnerUK) [inline tests/widget-core/unit/d.ts]: Does this need deleting?

File: src/widget-core/README.md [hunk 5]
```diff
@@ -130,17 +132,13 @@ class App extends WidgetBase {
...
 
-We can now use `App` with the `ProjectorMixin` to render the `Hello` widgets.
+We can now use `App` with the `renderer` to display the `Hello` widgets.
 
 ```ts
-const Projector = ProjectorMixin(App);
-const projector = new Projector();
-
-projector.append();
+const r = renderer(() => w(App, {}));
```
Comment (JamesMilnerUK) [inline src/widget-core/README.md]: This is much more straightforward and easy to understand. Nice :smile:

File: tests/widget-core/unit/d.ts (hunk 4 above)
Comment (agubler) [reply tests/widget-core/unit/d.ts]: Thanks, needs un-commenting!

File: src/widget-core/WidgetBase.ts [hunk 6]
```diff
@@ -261,19 +256,51 @@ export class WidgetBase<P = WidgetProperties, C extends DNode = DNode> implement
...
+	private _filterAndConvert(nodes: DNode[]): (WNode | VNode)[];
+	private _filterAndConvert(nodes: DNode): WNode | VNode;
+	private _filterAndConvert(nodes: DNode | DNode[]): (WNode | VNode) | (WNode | VNode)[];
+	private _filterAndConvert(nodes: DNode | DNode[]): (WNode | VNode) | (WNode | VNode)[] {
+		const isArray = Array.isArray(nodes);
+		const filteredNodes = Array.isArray(nodes) ? nodes : [nodes];
+		const convertedNodes: (WNode | VNode)[] = [];
+		for (let i = 0; i < filteredNodes.length; i++) {
+			const node = filteredNodes[i];
+			if (!node) {
```
[the remark this answers is not in the document]
Comment (JamesMilnerUK) [reply src/widget-core/WidgetBase.ts]: Right!

File: src/widget-core/WidgetBase.ts (hunk 2 above)
Comment (JamesMilnerUK) [reply src/widget-core/WidgetBase.ts]: Shame!

File: src/widget-core/WidgetBase.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (JamesMilnerUK) [reply src/widget-core/WidgetBase.ts]: Fair enough, more just an open ended question. If we don't spend much processing time here would imagine it won't matter much as you say.

[the remark this answers is not in the document]
Comment (agubler) [reply src/widget-core/WidgetBase.ts]: It is a hot path, runs for the render result of every widget being processed. I can certainly look at unrolling the loop and see if it makes any meaningful different.

**Comment by agubler** [comment]
Tested with js-framework-benchmark and examples - hnpwa, realworld, custom-element-menu, widget-showcase & custom-element-showcase.

**Pull request merged**